pkg/buildpack/buildpack.go

```diff
@@ -326,6 +356,133 @@ func ToLayerTar(dest string, module BuildModule) (string, error) {
 	return layerTar, nil
 }
 
+func ToNLayerTar(dest string, module BuildModule, logger logging.Logger) ([]ModuleTar, error) {
+	if !module.ContainsFlattenedModules() {
+		return handleSingleOrEmptyModule(dest, module)
+	}
+
+	modReader, err := module.Open()
+	if err != nil {
+		return nil, errors.Wrap(err, "opening blob")
+	}
+	defer modReader.Close()
+
+	tarCollection := newModuleTarCollection(dest)
+	rootFolderHeaders := map[string]*tar.Header{}
+
+	// the build module __could__ have more than one buildpack on it,
+	// if we detect multiple buildpacks, explode them
+	tr := tar.NewReader(modReader)
+	headersProcessed := 0
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				if headersProcessed == 0 {
+					// If the reader was empty, we need to write an empty tar on disk
+					return handleSingleOrEmptyModule(dest, module)
+				}
+				break
+			}
+			return nil, errors.Wrapf(err, "failed to read next data '%s'", header.Name)
+		}
+
+		id, version := parseBpIDAndVersion(header)
+		if !matchIDVersion(module.Descriptor(), id, version) {
+			if version == "" && header.Typeflag == tar.TypeDir {
+				// we need to save it and write it later
+				if _, ok := rootFolderHeaders[id]; !ok {
+					rootFolderHeaders[id] = header
+					continue
+				}
+			}
+		} else {
+			id = module.Descriptor().EscapedID()
+			version = module.Descriptor().Info().Version
+		}
+
+		bpTar, err := tarCollection.get(id, version)
+		if err != nil {
+			return nil, err
+		}
+
+		if rootHeader, ok := rootFolderHeaders[id]; ok {
+			err = bpTar.writer.WriteHeader(rootHeader)
+			logger.Debugf("writing root folder %s into module tar with id=%s", rootHeader.Name, id)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to write root header for '%s'", rootHeader.Name)
+			}
+			delete(rootFolderHeaders, id)
+		}
+
+		logger.Debugf("writing %s into module tar with id=%s, version=%s", header.Name, id, version)
+		err = bpTar.writer.WriteHeader(header)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to write header for '%s'", header.Name)
+		}
+
+		buf, err := io.ReadAll(tr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read contents of '%s'", header.Name)
+		}
+
+		_, err = bpTar.writer.Write(buf)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to write contents to '%s'", header.Name)
+		}
+		headersProcessed++
+	}
+
+	errs := tarCollection.close()
+	if len(errs) > 0 {
+		return nil, errors.New("closing files")
+	}
+
+	return tarCollection.getModules(), nil
+}
+
+func matchIDVersion(desc Descriptor, id, version string) bool {
+	sameID := id == desc.EscapedID() || id == desc.Info().ID
```

Comment:
I'm curious to know why we have to check both escaped and not escaped. Shouldn't we always see the escaped id when reading files from the package?

Reply:
I think I added that one because some unit tests, let me try removing it.